core/src/hash/hasher.rs

```diff
@@ -0,0 +1,360 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::{
+    hash::{
+        BuildHasherDefault,
+        Hasher,
+    },
+    marker::PhantomData,
+};
+use ink_prelude::vec::Vec;
+
+mod markers {
+    /// Prevents users from implementing certain traits or generic types.
+    pub trait Sealed {}
+
+    pub enum Sha2x256Marker {}
+    pub enum Keccakx256Marker {}
+    pub enum Blake2x256Marker {}
+    pub enum Blake2x128Marker {}
+    pub enum TwoxMarker {}
+
+    impl Sealed for Sha2x256Marker {}
+    impl Sealed for Keccakx256Marker {}
+    impl Sealed for Blake2x256Marker {}
+    impl Sealed for Blake2x128Marker {}
+    impl Sealed for TwoxMarker {}
+}
+
+/// Generic cryptographic hasher.
+pub struct CryptoHasher<T, A>
+where
+    T: markers::Sealed,
+{
+    /// The accumulator buffer.
+    ///
+    /// A bytes buffer that is used to accumulate a hashing state.
+    accumulator: A,
+    /// Tricks compiler into thinking that `Self` uses `T`.
+    marker: PhantomData<T>,
+}
+
+/// SHA2 256-bit hasher.
+pub type Sha2x256Hasher<I> = CryptoHasher<markers::Sha2x256Marker, I>;
+/// KECCAK 256-bit hasher.
+pub type Keccakx256Hasher<I> = CryptoHasher<markers::Keccakx256Marker, I>;
+/// BLAKE2 256-bit hasher.
+pub type Blake2x256Hasher<I> = CryptoHasher<markers::Blake2x256Marker, I>;
+/// BLAKE2 128-bit hasher.
+pub type Blake2x128Hasher<I> = CryptoHasher<markers::Blake2x128Marker, I>;
+/// TWOX 256, 128 or 64-bit hasher.
+pub type TwoxHasher<I> = CryptoHasher<markers::TwoxMarker, I>;
+
+/// Types that qualify as accumulator.
+///
+/// # Examples
+///
+/// Rust's `Vec<u8>` types and an exclusive reference to it (`&mut Vec<u8>`)
+/// qualify as such. Users may implement this trait for other types. E.g. it
+/// could be useful to have a `SmallVec` or a static buffer implementation for
+/// this trait.
+pub trait Accumulator {
+    /// Resets the buffer which cleans all state from it.
+    ///
+    /// # Note
+    ///
+    /// Useful when using `Vec` or similar as accumulator.
+    fn reset(&mut self);
+    /// Writes the given bytes into the buffer.
+    fn write(&mut self, bytes: &[u8]);
+    /// Returns a shared reference to the slice of the current state of the buffer.
+    fn as_slice(&self) -> &[u8];
+}
+
+impl Accumulator for Vec<u8> {
+    fn reset(&mut self) {
+        <Vec<_>>::clear(self)
+    }
+
+    fn write(&mut self, bytes: &[u8]) {
+        // This could theoretically be speed-up by using `unsafe` `set_len`
+        // and `[u8]` `copy_from_slice` methods.
+        <Vec<_>>::extend_from_slice(self, bytes)
```

Comment:
```suggestion
        self.extend_from_slice(bytes)
```

Reply:
I deliberately made this more explicit coming from a heavy macros background but in this situation you are right and we should apply your suggestions for the direct `Vec<u8>` implementations.